core/lib/dal/src/storage_logs_dal.rs

```diff
@@ -298,17 +246,23 @@ impl StorageLogsDal<'_, '_> {
             r#"
             SELECT DISTINCT
                 ON (hashed_key) hashed_key,
-                miniblock_number,
-                value
+                storage_logs.miniblock_number,
+                storage_logs.value
             FROM
                 storage_logs
+                LEFT JOIN miniblocks ON miniblocks.number = storage_logs.miniblock_number
+                LEFT JOIN snapshot_recovery ON snapshot_recovery.miniblock_number = storage_logs.miniblock_number
             WHERE
-                hashed_key = ANY ($1)
-                AND miniblock_number <= $2
+                storage_logs.hashed_key = ANY ($1)
+                AND storage_logs.miniblock_number <= $2
+                AND (
+                    miniblocks.number IS NOT NULL
+                    OR snapshot_recovery.miniblock_number IS NOT NULL
+                )
```

Comment:
`storage_logs.miniblock_number <= (SELECT MAX(number) FROM miniblocks)` won't work because some queries that have this construction are used in state keeper and miniblocks table is empt right after snapshot recovery. However, I can change it to smth like 
```sql
storage_logs.miniblock_number <= (SELECT MAX(number) FROM miniblocks)
OR
storage_logs.miniblock_number <= (SELECT miniblock_number FROM snapshot_recovery)
```
Do you think it will be better?

Reply:
Slightly better, but still doesn't look very good. I guess it could be possible to use `COALESCE` to collapse both sub-`SELECT`s, but it still would look messy.